lib/networkx/digraph.rb

```diff
@@ -0,0 +1,218 @@
+
+    # Constructor for initializing graph
+    #
+    # @example Initialize a graph with attributes 'type' and 'name'
+    #   graph = NetworkX::Graph.new(name: "Social Network", type: "undirected")
+    #
+    # @param graph_attrs [Hash{ Object => Object }] the graph attributes in a hash format
+    def initialize(**graph_attrs)
+      super(graph_attrs)
+
+      @pred = {}
+    end
+
+    # Adds the respective edge
+    #
+    # @example Add an edge with attribute name
+    #   graph.add_edge(node1, node2, name: "Edge1")
+    #
+    # @example Add an edge with no attribute
+    #   graph.add_edge("Bangalore", "Chennai")
+    #
+    # @param node_1 [Object] the first node of the edge
+    # @param node_2 [Object] the second node of the edge
+    # @param edge_attrs [Hash{ Object => Object }] the hash of the edge attributes
+    def add_edge(node_1, node_2, **edge_attrs)
+      add_node(node_1)
+      add_node(node_2)
+
+      edge_attrs = (@adj[node_1][node_2] || {}).merge(edge_attrs)
+      @adj[node_1][node_2] = edge_attrs
+      @pred[node_2][node_1] = edge_attrs
+    end
+
+    # Adds a node and its attributes to the graph
+    #
+    # @example Add a node with attribute 'type'
+    #   graph.add_node("Noida", type: "city")
+    #
+    # @param node [Object] the node object
+    # @param node_attrs [Hash{ Object => Object }] the hash of the attributes of the node
+    def add_node(node, **node_attrs)
+      super(node, node_attrs)
+
+      @pred[node] = {} unless @pred.key?(node)
+    end
+
+    # Removes node from the graph
+    #
+    # @example
+    #   graph.remove_node("Noida")
+    #
+    # @param node [Object] the node to be removed
+    def remove_node(node)
+      raise KeyError, "Error in deleting node #{node} from Graph." unless @nodes.key?(node)
+
+      neighbours = @adj[node]
+      neighbours.each_key { |k| @pred[k].delete(node) }
+      @pred[node].each_key do |k|
+        @adj[k].delete(node)
+      end
+
+      @pred.delete(node)
+      @adj.delete(node)
+      @nodes.delete(node)
+    end
+
+    # Removes edge from the graph
+    #
+    # @example
+    #   graph.remove_edge('Noida', 'Bangalore')
+    #
+    # @param node_1 [Object] the first node of the edge
+    # @param node_2 [Object] the second node of the edge
+    def remove_edge(node_1, node_2)
+      raise KeyError, "#{node_1} is not a valid node." unless @nodes.key?(node_1)
+      raise KeyError, "#{node_2} is not a valid node" unless @nodes.key?(node_2)
+      raise KeyError, 'The given edge is not a valid one.' unless @adj[node_1].key?(node_2)
+
+      @adj[node_1].delete(node_2)
+      @pred[node_2].delete(node_1)
+    end
+
+    # Clears the graph
+    #
+    # @example
+    #   graph.clear
+    def clear
+      super
+
+      @pred.clear
+    end
+
+    # Returns number of edges
+    #
+    # @example
+    #   graph.number_of_edges
+    def number_of_edges
+      num = 0
+      @adj.each { |_, v| num += v.length }
+      num
```

Comment:
Not sure if how this is better performance-wise, but you can probably also try it out.

```ruby
def number_of_edges
  @adj.values.map(&:length).inject(:+)
end
```